core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,288 @@
+import org.apache.hc.core5.http.Method;
+import org.apache.hc.core5.http.ParseException;
+import org.apache.hc.core5.http.impl.EnglishReasonPhraseCatalog;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.iceberg.exceptions.RESTException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.rest.responses.ErrorResponseParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An HttpClient for usage with the REST catalog.
+ */
+public class HttpRESTClient implements RESTClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpRESTClient.class);
+
+  private final String uri;
+  private final CloseableHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private final Map<String, String> additionalHeaders;
+  private final Consumer<HttpUriRequest> requestInterceptor;
+
+  private HttpRESTClient(
+      String uri,
+      CloseableHttpClient httpClient,
+      ObjectMapper mapper,
+      Map<String, String> additionalHeaders,
+      Consumer<HttpUriRequest> requestInterceptor) {
+    this.uri = uri;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : ImmutableMap.of();
+    this.requestInterceptor = requestInterceptor;
+  }
+
+  private static Optional<String> extractResponseBodyAsString(CloseableHttpResponse response, Class<?> responseType) {
+    try {
+      if (response.getEntity() == null) {
+        return Optional.empty();
+      }
+      return Optional.ofNullable(EntityUtils.toString(response.getEntity(), "UTF-8"));
+    } catch (IOException | ParseException e) {
+      if (responseType != null) {
+        LOG.warn("Failed to convert HTTP response body to string", e);
+      }
+      return Optional.empty();
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "ClientGeneratedErrorResponse - No error response body could be parsed";
```

Comment:
This doesn't look like a type. Why not use `RESTException` or something? The reason phrase should be enough right?

Reply:
RESTException is a good one. I wasn't really sure what to use, but the reason phrase should be plenty.

Reply:
This was updated to `RESTException` as the message is definitely sufficient.